### ISSUE TITLE
Multiple issues with WCF communications to stateful service

### ISSUE THREAD

**Issue by framgeld** [Azure/service-fabric-issues#867]
Hi,

Any assistance that anyone can provide on this will be VERY gratefully appreciated.

The issues I am experiencing are as follows (all on the dev cluster, 5 node mode), using the latest stable release of the Service Fabric NuGet packages (everything is Revision 456), and I have a feeling that they are all related (hence a single issue reported):

1. I must "yield" my custom WCF listener first in the **CreateServiceReplicaListeners()** method. If I do not do this, I get an exception in the WCF client that informs me that the "localhost" scheme is not valid, and that it should be "http". This is occurring because the **ServicePartitionResolver** is returning only one endpoint, containing both the ServiceProxy v2 listener address AND my WCF listener address. I believe I am doing something wrong, and would like to correct this so that I can publish my listeners in any order.
2. The WCF client (ASP.NET website, C# server-side) that is attempting to connect to my stateful service hangs when attempting to make the connection... **on the second attempt**. The first attempt works perfectly. I have narrowed this hang to the **ServicePartitionResolver.ResolveAsync** method. This could be because of point 3, below, but in the rare occurrences where my stateful service did not crash immediately, the **ResolveAsync** method still hung.
3. After I invoke the WCF client the first time, Service Fabric goes into a tailspin. I get reports that I have full quorum loss, and my stateful service crashes. Service Fabric attempts to restart my stateful service but it crashes (primary AND secondaries). In addition to this, Service Fabric will not allow me to delete the application instance, nor will it allow me to republish from Visual Studio 2017. I have to reset my entire dev cluster to rectify the issue so that I can attempt debugging again. I have debugged my WCF call within Service Fabric and it terminates without complaint, and my service starts / functions as expected until the WCF method is invoked. The tailspin appears to coincide with a warning that Service Fabric posts immediately after the WCF method has terminated: **The API IReplicator.CatchUpReplicaset is stuck**, and Service Fabric never recovers from it. My WCF method adds items to the Reliable Collections.

The service proxy v2 listener I refer to above was published as follows:

> localhost:30003+fe398b07-c679-42d1-8e57-7a7fb81406db-131636393301353633-93ab8673-4585-4ac5-8bf4-fff2e06d4e8c

My custom WCF listener was published as follows:
> http://localhost:30001/

The website acting as the WCF client is ASP.NET/MVC/.NET Framework 4.5/x64, and uses the following code to make the connection:

```
string[] serviceFabricClusterAddresses =
    {
        "localhost:19000",
        "localhost:19001",
    };
var serviceResolver = new ServicePartitionResolver(serviceFabricClusterAddresses);
var clientFactory =
    new WcfCommunicationClientFactory<IProducerWcfService>(
        GetBasicHttpBinding(),
        servicePartitionResolver: serviceResolver);
var uri = new Uri("fabric:/AppName/ServiceName");

return new ProducerWcfServiceClient(clientFactory, uri);

BasicHttpBinding GetBasicHttpBinding()
{
    const int OneMegabyteInBytes = 1024 * 1024;

    return
        new BasicHttpBinding
        {
            SendTimeout = TimeSpan.MaxValue,
            ReceiveTimeout = TimeSpan.MaxValue,
            OpenTimeout = TimeSpan.FromSeconds(5),
            CloseTimeout = TimeSpan.FromSeconds(5),
            MaxReceivedMessageSize = OneMegabyteInBytes,
            MaxBufferSize = OneMegabyteInBytes,
            MaxBufferPoolSize = OneMegabyteInBytes * Environment.ProcessorCount,
        };
}

```

The listeners are published as follows:

```
protected override IEnumerable<ServiceReplicaListener> CreateServiceReplicaListeners()
{
    yield return CreateProducerWcfServiceListener();

    foreach (var listener in this.CreateServiceRemotingReplicaListeners())
    {
        yield return listener;
    }

    ServiceReplicaListener CreateProducerWcfServiceListener()
    {
        const string ExternalWcfServiceEndpointResourceName = "ExternalWcfServiceEndpoint";
        return
            new ServiceReplicaListener(
                GetWcfCommunicationListener,
                ExternalWcfServiceEndpointResourceName);

        WcfCommunicationListener<IProducerWcfService> GetWcfCommunicationListener(
            StatefulServiceContext context)
        {
            string uri = GetUri();
            var listener =
                new WcfCommunicationListener<IProducerWcfService>(
                    context,
                    new ProducerWcfService(),
                    new BasicHttpBinding(BasicHttpSecurityMode.None),
                    new EndpointAddress(uri));

            return listener;

            string GetUri()
            {
                var endpointConfig =
                    context.CodePackageActivationContext.GetEndpoint(
                        ExternalWcfServiceEndpointResourceName);
                int port = endpointConfig.Port;
                string scheme = endpointConfig.Protocol.ToString();
                string host = context.NodeContext.IPAddressOrFQDN;
                return $"{scheme}://{host}:{port.ToString(CultureInfo.InvariantCulture)}/";
            }
        }
    }
}
```

The **ServiceManifest.xml** file contains the following endpoint definitions:

```
<Resources>
  <Endpoints>
    <Endpoint Name="ExternalWcfServiceEndpoint" Protocol="http" UriScheme="http" />
    <Endpoint Name="ServiceEndpointV2" />
    <Endpoint Name="ReplicatorEndpoint" />
  </Endpoints>
</Resources>
```

Thank you in advance.

**Comment by framgeld** [comment]
Further the to the issues above, I changed my dev cluster to run in single node mode to see what happens. My stateful service does not crash, and Service Fabric shows no errors or warnings. In addition, the WCF listener and the ServiceProxy listener remain responsive and function as expected.

When I switch it back to 5-node mode, as soon as I call the WCF service the first time, the following warning is raised, followed closely by all the stateful service replicas failing, crashing, and not being able to restart:

> Unhealthy event: SourceId='System.RAP', Property='IReplicator.CatchupReplicaSetDuration', HealthState='Warning', ConsiderWarningAsError=false.
> The api IReplicator.CatchupReplicaSet on node _Node_2 is stuck. Start Time (UTC): 2018-02-21 01:38:07.584.

Could there be something going on with the Reliable Collections replication?

**Comment by framgeld** [comment]
I have logged a separate issue for Point 2 in my original post (Points 1 and 3 still require assistance).

Issue https://github.com/Azure/service-fabric-issues/issues/874

**Comment by framgeld** [comment]
Further to point 3 in my original post: the Reliable Queues work fine until the first WCF call is made from my website. Once Service Fabric goes into a tailspin, attempting to enqueue further items results in the following exception:

> System.TimeoutException: Cannot backoff as reached max retry delay. MaxTotalDelay -2147483648 : Time Elapsed 16 : LastDelay 16 ---> System.Fabric.FabricTransientException: The operation failed because a quorum of nodes are not available for this replica set.  Consider retrying once more nodes are up. ---> System.Runtime.InteropServices.COMException: Exception from HRESULT: 0x80071BC5
>    at System.Fabric.Interop.NativeRuntime.IFabricStateReplicator2.EndReplicate(IFabricAsyncOperationContext context)
>    at System.Fabric.Interop.AsyncCallOutAdapter2`1.Finish(IFabricAsyncOperationContext context, Boolean expectedCompletedSynchronously)
>    --- End of inner exception stack trace ---
>    at Microsoft.ServiceFabric.Replicator.Transaction.ThrowReplicationException(Exception e, Boolean abortNeeded)
>    at Microsoft.ServiceFabric.Replicator.Transaction.PrivateAddOperation(OperationData metaData, OperationData undo, OperationData redo, Object operationContext, Int64 stateProviderId)
>    at Microsoft.ServiceFabric.Replicator.Transaction.<>c__DisplayClass11_0.<AddOperation>b__0()
>    at Microsoft.ServiceFabric.Replicator.TransactionStateMachine.OnAddOperation(Action addOperationAction)
>    at Microsoft.ServiceFabric.Data.ReplicationHelper.<ReplicateOperationAsync>d__0.MoveNext()
>    --- End of inner exception stack trace ---
>    at Microsoft.ServiceFabric.Data.RetryHelper.<ExponentialBackoffAsync>d__20.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.ServiceFabric.Data.ReplicationHelper.<ReplicateOperationAsync>d__0.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.ServiceFabric.Data.Collections.ReliableConcurrentQueue.ReliableConcurrentQueue`1.<ReplicateEnqueueAsync>d__57.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.ServiceFabric.Data.Collections.ReliableConcurrentQueue.ReliableConcurrentQueue`1.<Microsoft-ServiceFabric-Data-Collections-IReliableConcurrentQueue<T>-EnqueueAsync>d__54.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.GetResult()

**Comment by framgeld** [comment]
When the WCF call is made and Service Fabric starts going into meltdown, the events that are logged in Event Viewer are as follows:

> Warning: GetActivationContextAsyncOperation called for invalid HostType NonActivated

... followed by two occurrences of the following stack trace:

> Errormsg=1640c72a-20cf-446c-ac54-613d404ca1c5@131642864779303905@urn:Enqueued Jobs/dataStore@131642864875245229 Assert=System.Exception:    at System.Environment.GetStackTrace(Exception e, Boolean needFileInfo)
>    at System.Environment.get_StackTrace()
>    at System.Fabric.Store.TStore`5.OnApplyUpdate(TransactionBase txn, MetadataOperationData metadataOperationData, RedoUndoOperationData operationRedoUndo, Boolean isIdempotent, String applyType)
>    at System.Fabric.Store.TStore`5.<OnSecondaryApplyAsync>d__294.MoveNext()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
>    at System.Fabric.Store.TStore`5.OnSecondaryApplyAsync(Int64 sequenceNumber, TransactionBase replicatorTransaction, MetadataOperationData metadataOperationData, RedoUndoOperationData operationRedoUndo, Boolean isIdempotent)
>    at System.Fabric.Store.TStore`5.<Microsoft-ServiceFabric-Replicator-IStateProvider2-ApplyAsync>d__237.MoveNext()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
>    at System.Fabric.Store.TStore`5.Microsoft.ServiceFabric.Replicator.IStateProvider2.ApplyAsync(Int64 sequenceNumber, TransactionBase replicatorTransaction, OperationData metadata, OperationData operationData, ApplyContext applyContext)
>    at Microsoft.ServiceFabric.Replicator.DynamicStateManager.<OnApplyAsync>d__113.MoveNext()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
>    at Microsoft.ServiceFabric.Replicator.DynamicStateManager.OnApplyAsync(Int64 applyLSN, TransactionBase transaction, OperationData metaData, OperationData actualData, ApplyContext applyContext, Int64 stateProviderId)
>    at Microsoft.ServiceFabric.Replicator.DynamicStateManager.<OnApplyAsync>d__112.MoveNext()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
>    at Microsoft.ServiceFabric.Replicator.DynamicStateManager.OnApplyAsync(Int64 applyLSN, TransactionBase transaction, OperationData metaData, OperationData data, ApplyContext applyContext)
>    at Microsoft.ServiceFabric.Replicator.OperationProcessor.<ApplyCallback>d__36.MoveNext()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine](TStateMachine& stateMachine)
>    at Microsoft.ServiceFabric.Replicator.OperationProcessor.ApplyCallback(LogRecord record)
>    at Microsoft.ServiceFabric.Replicator.OperationProcessor.<ProcessLoggedRecordAsync>d__32.MoveNext()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine](TStateMachine& stateMachine)
>    at Microsoft.ServiceFabric.Replicator.OperationProcessor.ProcessLoggedRecordAsync(LogRecord record)
>    at Microsoft.ServiceFabric.Replicator.LogRecordsDispatcher.<ProcessTransaction>d__21.MoveNext()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine](TStateMachine& stateMachine)
>    at Microsoft.ServiceFabric.Replicator.LogRecordsDispatcher.ProcessTransaction(List`1 transaction)
>    at Microsoft.ServiceFabric.Replicator.LogRecordsDispatcher.<SeparateTransactions>d__22.MoveNext()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
>    at Microsoft.ServiceFabric.Replicator.LogRecordsDispatcher.SeparateTransactions()
>    at Microsoft.ServiceFabric.Replicator.LogRecordsDispatcher.<ProcessLoggedRecords>d__18.MoveNext()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
>    at Microsoft.ServiceFabric.Replicator.LogRecordsDispatcher.ProcessLoggedRecords()
>    at Microsoft.ServiceFabric.Replicator.LogRecordsDispatcher.<StartProcessingLoggedRecords>d__23.MoveNext()
>    at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine](TStateMachine& stateMachine)
>    at Microsoft.ServiceFabric.Replicator.LogRecordsDispatcher.StartProcessingLoggedRecords()
>    at Microsoft.ServiceFabric.Replicator.LogRecordsDispatcher.DispatchLoggedRecords(LoggedRecords loggedRecords)
>    at Microsoft.ServiceFabric.Replicator.LoggingReplicator.FlushedRecordsCallback(LoggedRecords loggedRecords)
>    at Microsoft.ServiceFabric.Replicator.PhysicalLogWriterCallbackManager.InvokeFlushedRecordsCallback(Object state)
>    at System.Threading.Tasks.Task.Execute()
>    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
>    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
>    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
>    at System.Threading.Tasks.Task.ExecuteEntry(Boolean bPreventDoubleExecution)
>    at System.Threading.ThreadPoolWorkQueue.Dispatch() at Time=‎2018‎-‎02‎-‎28T10:19:36.972825000Z

**Comment by zuhairp** [comment]
Hi @framgeld, can you share the traces from your local cluster? You can find the cluster traces fabric_traces_*.etl which is typically present at location %systemdrive%\SfDevCluster\Log\Traces.

**Comment by framgeld** [comment]
@BharatNarasimman This issue can now be closed. Thanks to everyone involved for their assistance.

For others experiencing the same issues, these were the causes:

1. The **ServicePartitionClient** constructor has an optional **listenerName** parameter. I was not setting this value.
2. This was related to async programming. The WCF client's WCF method was attempting to run asynchronously on the same ASP.NET worker thread that spawned the async Task for the WCF method call. As a result, it was causing deadlocks. I ensured my API "all the way down" was changed to use **async / await**, and the library **await** statements were changed to use **ConfigureAwait(false)**. This was logged separately on #874.
3. This was the trickiest of the three. The issues was related to the **GetHashCode** override method that I had created for my class, and the class was being used as a key in a Reliable Dictionary. My **GetHashCode** was performing a **.GetHashCode()** on standard .NET dictionaries (in my case, the dictionaries were empty). However, even though the dictionaries were empty, the dictionary does not yield the same hash code (even though they were "equivalently identical"). Service Fabric was trying to replicate to the active secondary replicas after the primary replica had performed a **TryUpdateAsync** on a Reliable Dictionary... but due to the hash code not generating in the same way, Service Fabric throws an exception because the dictionary key I provided was (for all intents and purposes) missing from the Reliable Dictionary.
